### DISCUSSION TITLE
Namespace Documentation feature

### DISCUSSION THREAD

**Discussion by a-michelis** [dotnet/docfx#7990] [Show and tell]
Even though, semantically, namespaces do not hold any value for intelisense (thus `<summary>` tags are not supported by namespaces), it is usually vital for documentations to provide explanatory texts in namespaces for both seamlessness and better understanding.

I currently am in a situation that requires such feature and I managed to offer it to myself by modifying the Metadata Extractor Worker with a wacky and possibly unstable workaround that kinda mimics `Sandcastle`'s approach.

Is there any plan for implementing ways of defining Namespace summaries?

**Comment by a-michelis** [comment]
# For anyone wondering what my solution was

On `\src\Microsoft.DocAsCode.Metadata.ManagedReference\ExtractMetadataWorker.cs`  i modified the method `GetMetadataFromProjectLevelCache` (and the class `ExtractMetadataWorker` in general) as follows:

**Before:**

```csharp
private Tuple<MetadataItem, bool> GetMetadataFromProjectLevelCache(IBuildController controller, IInputParameters key){
   // [...]
   projectMetadata = controller.ExtractMetadata(key);
   // [...]
}
```

**After:**

```csharp
private Tuple<MetadataItem, bool> GetMetadataFromProjectLevelCache(IBuildController controller, IInputParameters key){
   // [...]
   projectMetadata = controller.ExtractMetadata(key);
   ExtractNamespaceDocumentation(projectMetadata);
   // [...]
}

private void ExtractNamespaceDocumentation(MetadataItem node)
{
    // Terminal nodes are not of our interest in any case
    // Even if it's a namespace, it does not contain documentation
    if (node.Items is not { Count: > 0 }) return;
    
    // If it is namespace
    if (node.Type == MemberType.Namespace)
    {
        // Get (if any), the child that is class and is named "_NamespaceDoc"
        var doc = node.Items.FirstOrDefault(x =>
            x.Type == MemberType.Class && x.Name.Split('.').Last() == "_NamespaceDoc");

        // If we didn't found such class, the namespace does not contain documentation.
        // Leave and don't go further.
        if (doc is null) return;

        // Else, assign the class' Summary and remarks to the Namespace and remove the class from the tree.
        node.Summary = doc.Summary;
        node.Remarks = doc.Remarks;
        node.Items.Remove(doc);

        // job finished for this namespace, we do not want to go further down the tree.
        return;
    }

    // For non-namespace intermediate nodes (IE assembly nodes), visit the children.
    foreach (var child in node.Items) ExtractNamespaceDocumentation(child);
}
```

**My Project, at `\.namespaces.cs` :**

```csharp
namespace RootNamespace
{
    /// <summary>
    /// Documentation for the Root Namespace
    /// </summary>
    public static class _NamespaceDoc { }
}

namespace RootNamespace.SubFolder
{
    /// <summary>
    /// Documentation for the Root Namespace's `SubFolder` Sub-Namespace.
    /// </summary>
    /// <remarks>
    /// Remarks are also coppied...
    /// </remarks>
    public static class _NamespaceDoc { }
}

// [...]
```

**My `.csproj` :**

```XML
<!-- [...] -->
<ItemGroup Condition="'$(Configuration)'!='DOCUMENTATION'">
	<Compile Remove=".namespaces.cs" />
	<None Include=".namespaces.cs" />
</ItemGroup>
<!-- [...] -->
```

**My `docfx.json` :**

```json
{
  "metadata": [
    {
      "src": [
        {
          "files": [
            "**.csproj"
          ],
          "src": "/my/project/root"
        }
      ],
      "[...]" : "[...]",
      "properties": {
        "Configuration": "DOCUMENTATION"
      }
    }
  ],
  "[...]" : "[...]"
}
```

**Comment by KalleOlaviNiemitalo** [comment]
MicrosoftDocs has namespace summaries…

- [windows_foundation.md](https://github.com/MicrosoftDocs/winrt-api/blob/033412eebfde59331139f74a8dd9b14636a94904/windows.foundation/windows_foundation.md) is Markdown but contains special headings like `## -description`.
- [ns-System.xml](https://github.com/dotnet/dotnet-api-docs/blob/90e282640cd22936c129c9077447c8e710d2a0db/xml/ns-System.xml) is XML with embedded Markdown.

I don't know whether these are handled by DocFX or by some external tools. Neither of these matches the DocFX v3 [Markdown Fragments](https://github.com/dotnet/docfx/blob/a1fd171cae1740ba904c27da32c7dde8babbeab6/docs/designs/markdown-fragments.md) design document.

In DocFX v2, [Overwrite Files](https://github.com/dotnet/docfx/blob/76a2aaa560afe50e869c0a11451df0dbab26b3b7/Documentation/tutorial/intro_overwrite_files.md) might be usable for setting the summary of a namespace, but I don't remember trying that myself.

**Comment by a-michelis** [comment]
When trying to add summary to a namespace, autocompletion does not work (triple slash does not autocomplete any xml entity like `summary`) and DocFX v2 does not parse it either way. So, this:

![Namespace Summary](https://i.postimg.cc/X77gTTDj/ns1.png "Namespace Summary")

will result in this:

![Namespace Summary Result](https://i.postimg.cc/G244FxM1/ns2.png "Namespace Summary Result")

But my method:

![My Method](https://i.postimg.cc/8zykZrfk/ns3.png "My Method")

Will produce this:

![My Method Result](https://i.postimg.cc/GhHX4WPM/ns4.png "My Method Result")

Without including `_NamespaceDoc` class in both debugging or releasing

---

I do not know if namespace summary is actually supported (this way) by .NET (`<summary>` tags on namespaces are not included to the resulting XML file, upon building the project) and, in my understanding, that makes sense : What the compiler should do in case of a namespace having 2+ distinct summaries in different files?

I can see how my approach can be dangerous, tho.

--- 

About rewriting, on the other hand, since i tried it, i can confirm it works fine, but this approach does not unify code writting and documentation (which is what docfx tries to do) and does not provide any kind of intellisense (i cannot link to another documentation page unless i know the resulting HTML path and Heading ID) or even refactoring capabilities (if i change the name of my Namespace, i have to alter the `.md` file to point to the new Id and name).

**Reply by a-michelis** [reply]
Μy solution implies the premise that `_NamespaceDoc` is a class that, when defined, is used exclusively by docfx and this name cannot be changed, or used elsewere. But, adding this feature with additional switches* inside the `json` template could do the trick.

*by switches i mean something like

```json
{
  "metadata": [
    {
      "[...]" : "[...]",
      "namespace":{
        "enableDoc" : true,
        "className": "_NamespaceDoc"
      }
    }
  ],
  "[...]" : "[...]"
}
```

**Comment by cg3y** [comment]
Since V3 is now dead to the public, now would be a good time to focus on getting this added to DocFx. Have you considered adding a pull request to incorporate? This discussion seems pretty dead, yet this is an excellent feature addition (I need namespace documentation and it's cleaner to keep in it code in the triple slash format).

**Reply by paulushub** [reply]
[Overwrite Files](https://github.com/dotnet/docfx/blob/76a2aaa560afe50e869c0a11451df0dbab26b3b7/Documentation/tutorial/intro_overwrite_files.md) works well, why not move to that, which is more modern.

**Reply by a-michelis** [reply]
Sorry for being late, life happened!

It is my understanding that Overwrite Files defeats the purpose of inline documentation. It works? Yes. But writing the documentation in a completely different platform works as well.

Please correct me if i'm wrong, I'd love to try new features

**Comment by Patrick8639** [comment]
It's just a matter of adding an overwrite file for each namespace.
The overwrite file must have the following format:
```yaml
---
uid: <neplace with the namespace>
summary: *content
---

<Replace with the information about the namespace>
```

**Reply by a-michelis** [reply]
Will intelisense work on these documents as well? If yes, then that's, a totally viable option!

**Reply by Patrick8639** [reply]
It's a markdown file, with no relation to the source code, so IntelliSense is not working, unfortunately.

**Reply by paulushub** [reply]
IntelliSense for a namespace documentation? News to me, I have never used it.

**Reply by Patrick8639** [reply]
@paulushub I think he will make a summary of the available classes, or something similar.

**Reply by paulushub** [reply]
I understand what he is trying to do, but the point of IntelliSense he is raising seems off.
By the way, DocFX already provides a summary of classes available in a namespace. 
The overwrite files only provide a description of the namespaces, which is combined on the main namespaces page.
**_NamespaceDoc** was a hack to provide this namespace summary, which adds an empty class and it is not new - should we really do that?

![image](https://github.com/user-attachments/assets/c54d90e8-ddde-40e8-8948-d3e1f50770a5)

**Reply by a-michelis** [reply]
The point of intellisense is to provide auto completion, coherence in referencing other entities etc. 

Imagine refactoring your code and placing a class in a different namespace -or simply renaming the class or a field. This would require changing all the override documents you've been using so far, to point to this (new) reference. Using intelligence to refactor namespaces and namings in general, would also change the reference automatically.

I can understand your point if you write the documentation after you finished a project that's bound to a structure, but when you document code while implementing it or even if you're making dogfood versions of your software to be internally tested, this poses is a real issue -having to manually take time and review and rework all your documentation part of which would normally stay the same, (in case of simple moves or name changes)

As for the existing summary capabilities, if you divide call subsystems under different namespaces (which tends to be the case with c#) namespace summary is an extremely good place to write usage guides (ie: "how to use the audio subsystem in my game engine", under `MyGameEngine.Audio`)

**Reply by paulushub** [reply]
> The point of intellisense os to provide auto completion, coherence in referencing other entities etc.

The topic is **Namespace Documentation feature**, auto-completion has nothing to do with this as you rightly stated from the beginning 
```
Even though, semantically, namespaces do not hold any value for intelisense 
```
> Imagine refactoring your code and placing a class in a different namespace. 

No, you should not document the classes in the overwrite. Only the namespaces, the classes on the page above are automatically generated by DocFX.

**Reply by a-michelis** [reply]
I've edited my response to further strengthen my case.

> The topic is **Namespace Documentation feature**, auto-completion has nothing to do with this as you rightly stated from the beginning

The reason for having "in-code" namespace documentation insead of simply using overwrite files (which were there when the issue was made) IS to incorporate juicy, semantically-coupled information -see usage guides which could reference multiple sub-classes of a namespace. 

As the feature existed back then, why making an issue in the first place?

> No, you should not document the classes in the overwrite. 

I never wanted to document the classes themselves, but how various parts of the namespace can cooperate and be used in a system (or sub-system) level. And this is a completely valid approach, which, imho, is more efficient than dividing usage guide and documentation.

**Reply by paulushub** [reply]
> IS to incorporate juicy, semantically-coupled information

I will think this is hardly an issue, as the DocFX extensions to markdown make any such link possible.
Also, the documentation/DocFX project is part of your project just like unit tests.

> As the feature existed back then, why making an issue in the first place?

Because, it was a hack, not a .NET feature or feature extension. Using a class to document a namespace is **ODD**, not something many really wanted to do, and having to change the project file to specifically support this is not what many will opt for.
You are dealing with existing projects, that is understandable. Simply write a Python script to help your migration.

If what you want is something similar to Java's `package-info.java`,  then consider something like `namespacedoc` tag support and argue it out with the Roslyn and/or C# Team(s). Java's support does not use a class, and XML documents are extracted by the C# compiler.
```cs
using System;

/// <namespacedoc name="LabTests">
/// Your contents here...
/// </namespacedoc>
namespace LabTests;
```

As pointer out by the links above, many big company will rather document all sources in external file living only developer documentations (`devdoc`) in the codes - such teams have dedicated documentation teams. 
Previously, I think Microsoft was doing this
```cs
/// <devdoc>This is an interesting Foo class</devdoc>
/// <include file="docs.xml" path='Doc/Member[@for="MyNamespace.Foo"]/*' />
public class SomeClass { ... }
```
And they created the overwrite to support this type of flow.
You can see the documentation of `Calendar` class in `ICU4J`. Many companies will not want that per policy.
https://javadoc.io/doc/com.ibm.icu/icu4j/latest/index.html

**Reply by a-michelis** [reply]
The issue was created by me, back then. What I proposed is a workaround, not an actual solution to the problem. 

Your suggestion about the `namespacedoc` tag, would be awesome, and it might be worth it giving it a try! that's what I initially thought back then, but, unfortunately, it seems like Roslyn was not passing triple slash documentation to an XML parsing procedure, but rather supported specific tags with XML like structure. 

2 years ago I was way less knowledgeable, and I thought my case was a niche which only few would want, so I simply proposed the workaround to help those few people.

Now that my experience increases, tho, I find this feature all the more appealing, to the point that I return to this issue today. That said, I'm not proposing my workaround anymore, but rather want this feature to be implemented correctly -like your proposition does. 

As for big corpo, #c is a language that is not only used by big corporations: I for once have made countless tools for either internal or small volume consumption. These tools, being maintained by a very small group of developers or sometimes even only by me, would greatly benefit by such a feature. And I strongly believe the open source community that uses c# on smaller projects maintained by small group of people would really love this as well.

**Reply by Patrick8639** [reply]
> unfortunately, it seems like Roslyn was not passing triple slash documentation to an XML parsing procedure, but rather supported specific tags with XML like structure.

Roslyn also support any tag, as I've written a tool to post-process the generated XML file in order to support `<if>` conditions in the comments.
The problem is that we can't add XML documentation in a namespace declaration and thus, it doesn't export it.

A solution would perhaps be to define a class to hold all the namespace documentation (with one member for each namespace). The XML comment would contain `<namespacedoc>` and a post-processor would manage this to extract the namespace documentation.
To avoid having the class in the output assembly, we can can use `#if` to generate only when needed.

```cs
#if NamespaceDoc
sealed class NamespaceDoc
{
 /// <namespacedoc>
 /// Your contents here...
 /// </namespacedoc>
  private const Int32 LabTests = 0;
}
#endif
```

The information appears in the generated XML file and can be processed using a post-processor:
```xml
        <member name="F:xxx.NamespaceDoc.LabTests">
            <namespacedoc>
            Your contents here...
            </namespacedoc>
        </member>
```

**Reply by a-michelis** [reply]
In a way, that's what the workaround proposes. Not in the same sealed class but with different classes, but the logic is the same.

Also, yes, Roslyn **is** capable of parsing xml comments. What I meant is that its design does not foresee xml comments in namespace and thus doesn't attempt to retrieve/parse them.

**Comment by paulushub** [comment]
@a-michelis How about keeping this simple?
1. Add Readme.md to each namespace containing folder (so that anyone can read that and know what the namespace does)
2. You can pack these Readme files into overwrite files with scripts
3. If possible, add a pre-commit script or a git filter script to create the overwrite files.

I think this will require less changes, if any, to the project files, and can be part of the workflow.

**Reply by a-michelis** [reply]
That defeats whole in-code documentation purpose, in addition to increasing complexity, introducing custom procedures which may or may not work for every environment (ie windows vs linux users using scripts) and cause confusion to future maintainers.

Of course, you'll say "Ok, how often would you possibly wanna change namespace documentation", and yes, the answer is "not really that often". That said, it's not uncommon for procedures that are rarely initiated to be forgotten, and it's a buzkill for me. Sure, the project wont change for the likes of one quirky fella, but that's just an opinion that may help with the future of the project.

**Reply by paulushub** [reply]
> That defeats whole in-code documentation purpose

As to whether namespace documentation is considered in-code document is a view that many may or not agree.
For such, you can test your view by raising it up where it will actually be implemented, and see the reactions: 
Roslyn team in this case.

>  introducing custom procedures which may or may not work for every environment

Modern day development is a complex process, and nothing is fixed. Working with scripts to navigate
the complexity is one way to make life easier for oneself.  When things change you just update your scripts.
And there is a huge job category for just that: DevOps and its related processes.
Python/JavaScript (Node-js) are now basic part of most projects, and work across platform.

**Comment by Atrejoe** [comment]
Why not adhere to the solution Sandcastle has been using:
https://ewsoftware.github.io/SHFB/html/48f5a893-acde-4e50-8c17-72b83d9c3f9d.htm

``` c#
namespace Company.Product.Widgets
{
    /// <summary>
    /// These are the namespace comments for <c>Company.Product.Widgets</c>.
    /// </summary>
    [System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
    class NamespaceDoc
    {
    }
}
```

Any content is still compiler-validated.

**Reply by Atrejoe** [reply]
(sorry for waking up the dead, I just started exploring docfx, coming from Sandcastle.)

**Reply by a-michelis** [reply]
That's what my workaround essentially proposes

**Comment by calacayir** [comment]
FYI, this feature is implemented in my project [docfx-plus](https://github.com/dotmake-build/docfx-plus) which is a wrapper around docfx tool, which at runtime patches the internals to fix some problems and add new features.

- **Added:** `NamespaceDoc` support just like `SHFB`.  
  Namespace comments can be specified and maintained in your source code by adding an empty `NamespaceDoc` class to each namespace.  
  XML Comments (`<summary>` and `<remarks>` tags) from these classes will be extracted and assigned to the containing namespaces in the docs.
  
  ```c#
  namespace DotMake.CommandLine
  {
      /// <summary>
      /// This is the root namespace of this library, it includes the CLI attributes and CLI parser/runner.
      /// <para />
      /// <para>The main classes in this namespace are:</para>
      /// <list type="bullet">
      ///     <item><see cref="CliCommandAttribute"/> is the attribute which specifies a class that represents a command which is a specific action that the command line application performs.</item>
      ///     <item><see cref="CliOptionAttribute"/> is the attribute which specifies a class property that represents an option which is a named parameter and a value for that parameter, that is used on the command line.</item>
      /// </list>
      /// </summary>
      [System.Runtime.CompilerServices.CompilerGenerated]
      internal class NamespaceDoc
      {
      }
  }
  ```
  
  `NamespaceDoc` classes will be found even if your API filter excludes them but themselves will not be included in the docs,
  only their XML Comments assigned to the containing namespaces will be displayed.  
  You can optionally make them private or internal and mark them with a `CompilerGenerated` attribute.
